Sources/NIO/SelectorEpoll.swift

```diff
@@ -93,6 +93,35 @@ extension SelectorEventSet {
     }
 }
 
+@usableFromInline struct EPollUserData {
+    // We're backing this by a tuple to get standard C layout
+    @usableFromInline var _backing: (registrationID: SelectorRegistrationID, fileDescriptor: CInt)
+
+    @inlinable var registrationID: SelectorRegistrationID {
+        return self._backing.registrationID
+    }
+
+    @inlinable var fileDescriptor: CInt {
+        return self._backing.fileDescriptor
+    }
+
+    @inlinable init(registrationID: SelectorRegistrationID, fileDescriptor: CInt) {
+        assert(MemoryLayout<UInt64>.size == MemoryLayout<EPollUserData>.size)
+        self._backing = (registrationID, fileDescriptor)
+    }
+
+    @inlinable init(rawValue: UInt64) {
+        self._backing = unsafeBitCast(rawValue, to: type(of: self._backing))
```

Comment:
I don't love the `unsafeBitCast` here or below. We don't need it: the optimiser is quite capable of handling extracting the relevant fields to the appropriate layout. This seems to fall into a place where we're doing something unsafe unnecessarily, and it'll set off alarm bells every time I see it in the future.

Reply:
Hmm, `unsafeBitCast` gives us a guarantee that nothing else does: That `UInt64` and the backing has the _same size_. That's why I used it because I think it's actually more safe than doing some shifting/masking manually here. But yeah, it does have "Unsafe" in the name ;). Happy to add a comment though

Reply:
Or of course refactor to not use `unsafeBitCast` but IMHO this is kinda the right choice here?

Reply:
I don't agree that only `unsafeBitCast` can give us that guarantee. You wrote another line of code that can give that guarantee in the other `init`. 

But even if it were true I don't think that guarantee would be enough to justify the downsides. It makes this code much more dangerous to change in future, and that makes me very nervous.

Reply:
@Lukasa I missed a "at compile time" . uBC does it at compile time whereas `assert` only does it at runtime (in debug). And I need to manually shift/mask.

Reply:
uBC doesn't do it at compile time. Try out the following program:

```swift
func main() {
    let x = UInt64(5)
    let y = unsafeBitCast(x, to: UInt32.self)
    let a = UInt32(5)
    let b = unsafeBitCast(a, to: UInt64.self)
    print(y)
    print(b)
}

main()
```

This compiles just fine for me. It crashes at runtime, of course. As for "in debug", we can use `precondition` and get it in release mode too.

As for "you need to manually shift and mask": yes, you do. However, SSWG guidelines don't allow "I have to write more code" as a reason to deploy unsafe operations.

Reply:
Okay, I was convinced that `error: Can't unsafeBitCast between types of different sizes` is compile time but seems like you're right! Will replace in this case. We should file a bug for this though, will do later.